cmd/object-handlers.go

```diff
@@ -2913,26 +2990,11 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
 		scheduleReplicationDelete(ctx, dobj, objectAPI, replicateSync)
 	}
 
-	if goi.TransitionStatus == lifecycle.TransitionComplete { // clean up transitioned tier
-		// if version id being deleted, delete from transition tier
-		// if dm being set, noop on transition tier
-		// non versioned obj being deleted, delete from transition tier.
-		delTier := false
-		switch {
-		case opts.VersionID != "":
-			delTier = true
-		case !globalBucketVersioningSys.Enabled(bucket):
-			delTier = true
-		}
-		if delTier {
-			deleteTransitionedObject(ctx, newObjectLayerFn(), bucket, object, lifecycle.ObjectOpts{
-				Name:             object,
-				UserTags:         goi.UserTags,
-				VersionID:        goi.VersionID,
-				DeleteMarker:     goi.DeleteMarker,
-				TransitionStatus: goi.TransitionStatus,
-				IsLatest:         goi.IsLatest,
-			}, goi.transitionedObjName, goi.TransitionTier, false, false)
+	if hasLifecycleConfig {
+		// Remove the transitioned object whose object version is being overwritten.
+		if jentry, ok := os.ShouldRemoveRemoteObject(); ok {
```

Comment:
ShouldRemoveRemoteObject should not automatically delete the remote object if versioning is suspended on bucket - I think this attempt to unify can cause way more bugs if we are not careful